File: Data Structures & Algorithms/count-paths/submission-3.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
       
        memo = {}
        def dfs(i, j):
            # print(i, j)
            count =0
            if (i, j) in memo:
                return memo[(i, j)]
            if i == m-1 and j == n-1:
                return 1;
            elif i == m-1:  
                count+=dfs(i, j+1) 
            elif j==n-1: 
                count+=dfs(i+1, j)
            else:
                count+=dfs(i, j+1)
                count+=dfs(i+1, j)
            
            memo[(i, j)] = count
            return count

       
        return  dfs(0, 0)

File: Data Structures & Algorithms/word-break/submission-3.java
class Solution {
    Map<String, Boolean> memo = new HashMap<>();
    Boolean dfs(String s, List<String> wordDict) {
        if(s.length() == 0) {
            return true;
        }
        if (memo.containsKey(s)) {
            return memo.get(s);
        }
        for(String word: wordDict) {
            if (s.length() >= word.length() && s.substring(0, word.length()).equals(word)) {
                String unCheckedStr = s.substring(word.length(), s.length());
                
                if (dfs(unCheckedStr, wordDict)) {
                    memo.put(s, true);
                    return true;
                }

            }   
        }
        memo.put(s, false);
        return false;
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        return dfs(s, wordDict);
    }
}
